server/server.go

```diff
@@ -408,6 +408,178 @@ func (s *server) Start(ctx context.Context, req *rpcpb.StartRequest) (*rpcpb.Sta
 	return &rpcpb.StartResponse{ClusterInfo: s.clusterInfo}, nil
 }
 
+func (s *server) DeployBlockchains(ctx context.Context, req *rpcpb.DeployBlockchainsRequest) (*rpcpb.DeployBlockchainsResponse, error) {
+	// if timeout is too small or not set, default to 5-min
+	if deadline, ok := ctx.Deadline(); !ok || time.Until(deadline) < DefaultStartTimeout {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(context.Background(), DefaultStartTimeout)
+		_ = cancel // don't call since "start" is async, "curl" may not specify timeout
+		zap.L().Info("received start request with default timeout", zap.String("timeout", DefaultStartTimeout.String()))
+	} else {
+		zap.L().Info("received start request with existing timeout", zap.String("deadline", deadline.String()))
+	}
+
+	zap.L().Debug("DeployBlockchains")
+	if info := s.getClusterInfo(); info == nil {
+		return nil, ErrNotBootstrapped
+	}
+
+	chainSpecs := []blockchainSpec{}
+	// TODO: fix plugin dir management:
+	// - cannot change plugin dir for nodes that wont be restarted
+	// - can change plugin dir when restating network
+	// - different nodes may have different plugin dirs
+	// - pluging dir shall contain vms for all blockchains
+	if req.GetPluginDir() == "" {
+		if len(req.GetBlockchainSpecs()) > 0 {
+			return nil, ErrPluginDirEmptyButCustomVMsNotEmpty
+		}
+	} else {
+		if len(req.GetBlockchainSpecs()) == 0 {
+			return nil, ErrPluginDirNonEmptyButCustomVMsEmpty
+		}
+		zap.L().Info("non-empty plugin dir", zap.String("plugin-dir", req.GetPluginDir()))
+		for i := range req.GetBlockchainSpecs() {
+			vmName := req.GetBlockchainSpecs()[i].VmName
+			vmGenesisFilePath := req.GetBlockchainSpecs()[i].Genesis
+			zap.L().Info("checking custom VM ID before installation", zap.String("vm-id", vmName))
+			vmID, err := utils.VMID(vmName)
+			if err != nil {
+				zap.L().Warn("failed to convert VM name to VM ID",
+					zap.String("vm-name", vmName),
+					zap.Error(err),
+				)
+				return nil, ErrInvalidVMName
+			}
+			if err := utils.CheckPluginPaths(
+				filepath.Join(req.GetPluginDir(), vmID.String()),
+				vmGenesisFilePath,
+			); err != nil {
+				return nil, err
+			}
+			b, err := os.ReadFile(vmGenesisFilePath)
+			if err != nil {
+				return nil, err
+			}
+			chainSpecs = append(chainSpecs, blockchainSpec{
+				vmName:   vmName,
+				genesis:  b,
+				subnetId: req.GetBlockchainSpecs()[i].SubnetId,
+			})
+		}
+	}
+
+	s.mu.Lock()
```

Comment:
I wonder what these locks are actually doing now...It unlocks as soon as the function is over, but there is the async `go func` for listening until the `deployBlockchainsReadyCh` returns. But that means that during that time it is unlocked, so something can be changed before `deployBlockchainsReadyCh` is closed.

Could that create problems? This probably would apply to AddSubnets and start as well, but just mentioning here.

Reply:
This comment also applies to original Start() of the server, the mutex usage is the same as before.
I shall make another PR to work on that issue. 
The locks, before and now, are controlling two things:
- mutual exclusion between sync calls (which also includes the RemoveNode/AddNode/Waits inside subnet creation)
- mutual exclusion when updating server information from network information
They are not doing mutual exclusion between async and sync calls
I suggest to make another PR after merging this, that could give this semantics:

Mutual exclusion between server operations (except obtaining status and stopping) 
- convert sync calls into async ones
- mutual exclusion between all async call (old and new), using a mutex exclusive for that
- so obtaining queueing orders now for almost all server operations

Mutual exclusion for server information read/write
- mutual exclusion which a exclusive mutex, to upate server information from network information
- get state calls use same mutex

Stop operation to no mutex on server operations, having capability to stop any sync op

Reply:
https://ava-labs.atlassian.net/browse/AV-1761